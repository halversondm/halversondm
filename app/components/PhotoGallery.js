/**
 * Created by Daniel on 6/28/2016.
 */
"use strict";

import React from "react";

const PhotoGallery = React.createClass({

  propTypes: {
    perPage: React.PropTypes.number,
    totalPhotos: React.PropTypes.number,
    filePrefix: React.PropTypes.string,
    fileSuffix: React.PropTypes.string
  },
  getDefaultProps() {
    return {
      perPage: 0,
      totalPhotos: 0,
      filePrefix: "",
      fileSuffix: ""
    };
  },
  getInitialState() {
    return {
      photoArray: [],
      buttons: [],
      firstPhoto: [],
      lastPhoto: [],
      hidePrevious: false,
      hideNext: false
    };
  },
  componentDidMount() {
    this.init();
  },
  init() {
    var go = true;
    var page = 0;
    var firstPhoto = [];
    var lastPhoto = [];
    var buttons = [];
    while (go) {
      if (page === 0) {
        firstPhoto.push(1);
        lastPhoto.push(this.props.perPage);
      } else {
        var nextFirst = lastPhoto[page - 1] + 1;
        var nextLast = nextFirst + this.props.perPage;
        if (nextLast > this.props.totalPhotos) {
          nextLast = this.props.totalPhotos;
        }
        firstPhoto.push(nextFirst);
        lastPhoto.push(nextLast);
      }

      buttons.push(page + 1);
      if (lastPhoto[page] === this.props.totalPhotos) {
        go = false;
      } else {
        page += 1;
      }
    }
    this.setState({
      buttons: buttons,
      firstPhoto: firstPhoto,
      lastPhoto: lastPhoto
    });
    this.buildArray(1, firstPhoto, lastPhoto);
  },
  buildArray(pageNumber, firstPhoto, lastPhoto) {
    var photoArray = [];
    for (var i = firstPhoto[pageNumber - 1]; i <= lastPhoto[pageNumber - 1]; i += 1) {
      var source = this.props.filePrefix + i + this.props.fileSuffix;
      photoArray.push(source);
    }
    this.setState({photoArray: photoArray});
  },
  click(pageNumber) {
    this.buildArray(pageNumber, this.state.firstPhoto, this.state.lastPhoto);
  },
  clickAction(event) {
    var photoIndex = Number(event.currentTarget.dataset.i);
    this.showHideButtons(photoIndex);
  },
  prev() {
    var photoIndex = this.state.photoIndex - 1;
    this.showHideButtons(photoIndex);
  },
  next() {
    var photoIndex = this.state.photoIndex + 1;
    this.showHideButtons(photoIndex);
  },
  showHideButtons(photoIndex) {
    var hideNext = photoIndex === (this.state.photoArray.length - 1);
    var hidePrevious = photoIndex === 0;
    this.setState({
      hideNext: hideNext,
      hidePrevious: hidePrevious,
      photoIndex: photoIndex
    });
  },
  render() {
    return <div id="photoGallery">
      <ul className="row">
        {
          this.state.photoArray.map((photo, i) => {
            return <li key={i} className="col-lg-2 col-md-2 col-sm-3 col-xs-4">
              <img onClick={this.clickAction} data-i={i}
                   className="img-responsive" data-toggle="modal"
                   data-target="#photoModal"
                   src={photo}/>
            </li>;
          })
        }
      </ul>
      <div id="photoModal" className="modal">
        <div className="modal-dialog" role="document">
          <div className="modal-content">
            <div className="modal-body">
              <img src={this.state.photoArray[this.state.photoIndex]}
                   className="img-responsive"/>
            </div>
            <div className="modal-footer">
              <ul className="pager">
                <li className="previous">
                  <input type="button" hidden={this.state.hidePrevious}
                         onClick={this.prev} value="&larr; Previous"/>
                </li>
                <li className="next">
                  <input type="button" hidden={this.state.hideNext}
                         onClick={this.next} value="Next &rarr;"/>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>;
  }
});

export default PhotoGallery;
