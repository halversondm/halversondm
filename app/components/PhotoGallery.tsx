/**
 * Created by Daniel on 6/28/2016.
 */
import * as React from "react";
import {Button, Modal, Row} from "react-bootstrap";

export interface PhotoGalleryProps {
    perPage: number;
    totalPhotos: number;
    filePrefix: string;
    fileSuffix: string;
}

export interface PhotoGalleryState {
    photoArray: string[];
    pages: string[];
    firstPhoto: number[];
    lastPhoto: number[];
    hidePrevious: boolean;
    hideNext: boolean;
    showModal: boolean;
    photoIndex: number;
}

export class PhotoGallery extends React.Component<PhotoGalleryProps, PhotoGalleryState> {

    constructor(props: PhotoGalleryProps) {
        super(props);
        this.state = {
            photoArray: [],
            pages: [],
            firstPhoto: [],
            lastPhoto: [],
            hidePrevious: false,
            hideNext: false,
            showModal: false,
            photoIndex: 0,
        };
        this.init = this.init.bind(this);
        this.buildArray = this.buildArray.bind(this);
        this.imageClick = this.imageClick.bind(this);
        this.close = this.close.bind(this);
        this.next = this.next.bind(this);
        this.prev = this.prev.bind(this);
        this.pageClick = this.pageClick.bind(this);
        this.showHideButtons = this.showHideButtons.bind(this);
    }

    componentDidMount() {
        this.init();
    }

    init() {
        let go = true;
        let page = 0;
        const firstPhoto = [];
        const lastPhoto = [];
        const pages = [];
        while (go) {
            if (page === 0) {
                firstPhoto.push(1);
                lastPhoto.push(this.props.perPage);
            } else {
                const nextFirst = lastPhoto[page - 1] + 1;
                let nextLast = nextFirst + this.props.perPage;
                if (nextLast > this.props.totalPhotos) {
                    nextLast = this.props.totalPhotos;
                }
                firstPhoto.push(nextFirst);
                lastPhoto.push(nextLast);
            }

            pages.push(page + 1);
            if (lastPhoto[page] === this.props.totalPhotos) {
                go = false;
            } else {
                page += 1;
            }
        }
        this.setState({
            pages,
            firstPhoto,
            lastPhoto,
        });
        this.buildArray(1, firstPhoto, lastPhoto);
    }

    buildArray(pageNumber, firstPhoto, lastPhoto) {
        const photoArray = [];
        for (let i = firstPhoto[pageNumber - 1]; i <= lastPhoto[pageNumber - 1]; i += 1) {
            const source = this.props.filePrefix + i + this.props.fileSuffix;
            photoArray.push(source);
        }
        this.setState({photoArray});
    }

    imageClick(event) {
        const photoIndex = Number(event.currentTarget.dataset.i);
        this.showHideButtons(photoIndex);
        this.setState({showModal: true});
    }

    close() {
        this.setState({showModal: false});
    }

    prev() {
        const photoIndex = this.state.photoIndex - 1;
        this.showHideButtons(photoIndex);
    }

    next() {
        const photoIndex = this.state.photoIndex + 1;
        this.showHideButtons(photoIndex);
    }

    showHideButtons(photoIndex) {
        const hideNext = photoIndex === (this.state.photoArray.length - 1);
        const hidePrevious = photoIndex === 0;
        this.setState({
            hideNext,
            hidePrevious,
            photoIndex,
        });
    }

    pageClick(event) {
        const page = Number(event.currentTarget.dataset.id);
        this.buildArray(page, this.state.firstPhoto, this.state.lastPhoto);
    }

    render() {
        return (
            <div id="photoGallery">
                <ul className="row">
                    {
                        this.state.photoArray.map((photo, i) => {
                            return <li key={i} className="col-md-2 col-lg-2 col-sm-3 col-xs-4">
                                <img onClick={this.imageClick} data-i={i}
                                     className="img-fluid"
                                     src={photo}/>
                            </li>;
                        })
                    }
                </ul>
                <Row>
                    {
                        this.state.pages.map((page, i) => {
                            return <div key={i}><Button data-id={page} variant="primary"
                                                        onClick={this.pageClick}>{"Page " + page}</Button></div>;
                        })
                    }
                </Row>
                <Modal show={this.state.showModal} onHide={this.close}>
                    <Modal.Body>
                        <img src={this.state.photoArray[this.state.photoIndex]}
                             className="img-fluid"/>
                    </Modal.Body>
                    <Modal.Footer>
                        <ul className="pager">
                            <li className="previous">
                                <Button variant="primary" hidden={this.state.hidePrevious}
                                        onClick={this.prev}>&larr; Previous</Button>
                            </li>
                            <li className="next">
                                <Button variant="primary" hidden={this.state.hideNext}
                                        onClick={this.next}>Next &rarr;</Button>
                            </li>
                        </ul>
                    </Modal.Footer>
                </Modal>
            </div>
        );
    }
}
