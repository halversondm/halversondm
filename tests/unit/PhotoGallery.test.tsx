import React from "react";
import { render } from '@testing-library/react';
import PhotoGallery from "../../app/components/PhotoGallery";

test("Photo Gallery to match", () => {
  const unit = (
    <PhotoGallery
      perPage={5}
      filePrefix="photo"
      fileSuffix=".jpg"
      totalPhotos={10}
    />
  );
  const component = render(unit);
  const tree = component.container;
  expect(tree.toString()).toMatchSnapshot();
});
