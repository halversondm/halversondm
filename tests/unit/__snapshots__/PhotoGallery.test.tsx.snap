// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo Gallery to match 1`] = `
<div
  id="photoGallery"
>
  <ul
    className="row photo-gallery"
  >
    <li
      className="col-md-2 col-lg-2 col-sm-3 col-xs-4"
    >
      <img
        className="img-fluid"
        data-i={0}
        onClick={[Function]}
        src="photo1.jpg"
      />
    </li>
    <li
      className="col-md-2 col-lg-2 col-sm-3 col-xs-4"
    >
      <img
        className="img-fluid"
        data-i={1}
        onClick={[Function]}
        src="photo2.jpg"
      />
    </li>
    <li
      className="col-md-2 col-lg-2 col-sm-3 col-xs-4"
    >
      <img
        className="img-fluid"
        data-i={2}
        onClick={[Function]}
        src="photo3.jpg"
      />
    </li>
    <li
      className="col-md-2 col-lg-2 col-sm-3 col-xs-4"
    >
      <img
        className="img-fluid"
        data-i={3}
        onClick={[Function]}
        src="photo4.jpg"
      />
    </li>
    <li
      className="col-md-2 col-lg-2 col-sm-3 col-xs-4"
    >
      <img
        className="img-fluid"
        data-i={4}
        onClick={[Function]}
        src="photo5.jpg"
      />
    </li>
  </ul>
  <div
    className="row"
  >
    <div>
      <button
        className="btn btn-primary"
        data-id={1}
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Page 1
      </button>
    </div>
    <div>
      <button
        className="btn btn-primary"
        data-id={2}
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Page 2
      </button>
    </div>
  </div>
</div>
`;
